src/types/custom.rs

```diff
@@ -1,27 +1,35 @@
 //! Opaque types, used to represent a user-defined [`SimpleType`].
+//!
+//! [`SimpleType`]: super::SimpleType
 use smol_str::SmolStr;
 use std::fmt::{self, Display};
 
-use super::{ClassicType, SimpleType, TypeRow};
+use super::{type_param::TypeArg, ClassicType};
 
 /// An opaque type element. Contains an unique identifier and a reference to its definition.
 //
 // TODO: We could replace the `Box` with an `Arc` to reduce memory usage,
 // but it adds atomic ops and a serialization-deserialization roundtrip
 // would still generate copies.
-#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+#[derive(Debug, PartialEq, Eq, Clone, serde::Serialize, serde::Deserialize)]
 pub struct CustomType {
     /// Unique identifier of the opaque type.
-    id: SmolStr,
-    params: Box<TypeRow>,
+    /// Same as the corresponding [`TypeDef`]
+    ///
+    /// [`TypeDef`]: crate::resource::TypeDef
+    pub id: SmolStr,
+    /// Arguments that fit the [`TypeParam`]s declared by the typedef
+    ///
+    /// [`TypeParam`]: super::type_param::TypeParam
+    pub params: Vec<TypeArg>,
 }
 
 impl CustomType {
     /// Creates a new opaque type.
-    pub fn new(id: SmolStr, params: impl Into<TypeRow>) -> Self {
+    pub fn new(id: SmolStr, params: impl Into<Vec<TypeArg>>) -> Self {
         Self {
             id,
-            params: Box::new(params.into()),
+            params: params.into(),
         }
     }
```

Comment:
```rust
    pub fn new(id: impl Into<SmolStr>, params: impl Into<Vec<TypeArg>>) -> Self {
        Self {
            id: id.into(),
            params: params.into(),
        }
    }
```

and since it's a common case, it would be nice to have a const-constructor

```rust
    pub const fn new_simple(id: SmolStr) -> Self {
        Self {
            id,
            params: Vec::new(),
        }
    }
```

Reply:
Yes, nice :-) - done.